pkg/core/native/notary.go

```diff
@@ -0,0 +1,345 @@
+	prefixDeposit = 1
+	// notaryRewardFactor is a factor by which notary node reward per notaries is multiplied
+	// to make calculations more precise.
+	notaryRewardFactor = 100_000_000
+	// defaultDepositLockedHeight is a default chain's height until which deposit is locked.
+	defaultDepositLockedHeight uint32 = 5760 // TODO
+)
+
+// newNotary returns Notary native contract.
+func newNotary() *Notary {
+	n := &Notary{ContractMD: *interop.NewContractMD(notaryName)}
+	n.ContractID = notaryContractID
+
+	desc := newDescriptor("onPayment", smartcontract.VoidType,
+		manifest.NewParameter("from", smartcontract.Hash160Type),
+		manifest.NewParameter("amount", smartcontract.IntegerType))
+	md := newMethodAndPrice(n.onPayment, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("lockDepositUntil", smartcontract.BoolType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type),
+		manifest.NewParameter("till", smartcontract.IntegerType))
+	md = newMethodAndPrice(n.lockDepositUntil, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("withdraw", smartcontract.BoolType,
+		manifest.NewParameter("from", smartcontract.Hash160Type),
+		manifest.NewParameter("to", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.withdraw, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("balanceOf", smartcontract.IntegerType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.balanceOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("expirationOf", smartcontract.IntegerType,
+		manifest.NewParameter("addr", smartcontract.Hash160Type))
+	md = newMethodAndPrice(n.expirationOf, 100_0000, smartcontract.AllowStates)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("verify", smartcontract.BoolType)
+	md = newMethodAndPrice(n.verify, 100_0000, smartcontract.NoneFlag)
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("name", smartcontract.StringType)
+	md = newMethodAndPrice(nameMethod(notaryName), 0, smartcontract.NoneFlag)
+	n.AddMethod(md, desc, true)
+
+	desc = newDescriptor("onPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(n.OnPersist), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	desc = newDescriptor("postPersist", smartcontract.VoidType)
+	md = newMethodAndPrice(getOnPersistWrapper(postPersistBase), 0, smartcontract.AllowModifyStates)
+	n.AddMethod(md, desc, false)
+
+	return n
+}
+
+// Metadata implements Contract interface.
+func (n *Notary) Metadata() *interop.ContractMD {
+	return &n.ContractMD
+}
+
+// Initialize initializes Notary native contract and implements Contract interface.
+func (n *Notary) Initialize(ic *interop.Context) error {
+	return nil
+}
+
+// OnPersist implements Contract interface.
+func (n *Notary) OnPersist(ic *interop.Context) error {
```

Comment:
I've just noticed that we're actually missing one important piece in the original description of `OnPersist`, we must burn the deposit made for transactions where the sender is notary contract (`tx.NetworkFee + tx.SystemFee` amount, this will be burned by GAS contract from notary contract address, so we need to reflect that in our deposits).